Added PEFT docs (#1133)
huggingface/hub-docs: docs/hub/peft.md

> # Using PEFT at Hugging Face

Reviewer: Let's also add to https://huggingface.co/docs/hub/models-libraries

> 🤗 [Parameter-Efficient Fine-Tuning (PEFT)](https://huggingface.co/docs/peft/index), is a library for efficiently adapting pre-trained language models to various downstream applications without fine-tuning all the model’s parameters. 

Reviewer: ```suggestion
🤗 [Parameter-Efficient Fine-Tuning (PEFT)](https://huggingface.co/docs/peft/index) is a library for efficiently adapting pre-trained language models to various downstream applications without fine-tuning all the model’s parameters. 
```

## Exploring PEFT on the Hub

You can find PEFT models by filtering at the left of the [models page](https://huggingface.co/models?library=peft&sort=trending).

## Installation

To get started, you can check out the [Quick Tour of PEFT docs](https://huggingface.co/docs/peft/main/en/quicktour#quicktour). To install, follow the [PEFT installation guide](https://huggingface.co/docs/peft/install).
You can also use the following one-line install through pip:

```
$ pip install peft
```

## Using existing models

All PEFT models can be loaded from Hub. Loading PEFT models require loading base models that were fine-tuned, so we need to load the transformers model as well.

```py
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import PeftModel, PeftConfig

> config = PeftConfig.from_pretrained("dfurman/Mistral-7B-Instruct-v0.2")
> model = AutoModelForCausalLM.from_pretrained("mistralai/Mistral-7B-v0.1")
> model = PeftModel.from_pretrained(model, "dfurman/Mistral-7B-Instruct-v0.2")
> tokenizer = AutoTokenizer.from_pretrained(config.base_model_name_or_path)

Reviewer: Maybe this makes things a bit cleaner
```suggestion
base_model = "mistralai/Mistral-7B-v0.1"
adapter_model = "dfurman/Mistral-7B-Instruct-v0.2"

model = AutoModelForCausalLM.from_pretrained(base_model)
model = PeftModel.from_pretrained(model, adapter_model)
tokenizer = AutoTokenizer.from_pretrained(base_model)
```

model = model.to("cuda")
model.eval()
```

Once loaded, you can pass inputs to tokenizer to prepare them, and call `model.generate()` in regular PyTorch fashion.

```py
inputs = tokenizer("Tell me the recipe for chocolate chip cookie", return_tensors="pt")

with torch.no_grad():
    outputs = model.generate(input_ids=inputs["input_ids"].to("cuda"), max_new_tokens=10)
    print(tokenizer.batch_decode(outputs.detach().cpu().numpy(), skip_special_tokens=True)[0])
```

It outputs the following:

```text
Tell me the recipe for chocolate chip cookie dough.

> 1. Preheat oven

Reviewer: Maybe increase the number of generated tokens so we have a nicer example

# Using PEFT at Hugging Face

🤗 [Parameter-Efficient Fine-Tuning (PEFT)](https://huggingface.co/docs/peft/index) is a library for efficiently adapting pre-trained language models to various downstream applications without fine-tuning all the model’s parameters. 

## Exploring PEFT on the Hub

You can find PEFT models by filtering at the left of the [models page](https://huggingface.co/models?library=peft&sort=trending).

## Installation

> To get started, you can check out the [Quick Tour of PEFT docs](https://huggingface.co/docs/peft/main/en/quicktour#quicktour). To install, follow the [PEFT installation guide](https://huggingface.co/docs/peft/install).

Reviewer 2: ```suggestion
To get started, you can check out the [Quick Tour in the PEFT docs](https://huggingface.co/docs/peft/quicktour). To install, follow the [PEFT installation guide](https://huggingface.co/docs/peft/install).
```

You can also use the following one-line install through pip:

```
$ pip install peft
```

## Using existing models

> All PEFT models can be loaded from Hub. Loading PEFT models require loading base models that were fine-tuned, so we need to load the transformers model as well.

Reviewer 2: ```suggestion
All PEFT models can be loaded from the Hub. To use a PEFT model you also need to load the base model that was fine-tuned, as shown below.
```

```py
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import PeftModel, PeftConfig

base_model = "mistralai/Mistral-7B-v0.1"
adapter_model = "dfurman/Mistral-7B-Instruct-v0.2"

model = AutoModelForCausalLM.from_pretrained(base_model)
model = PeftModel.from_pretrained(model, adapter_model)
tokenizer = AutoTokenizer.from_pretrained(base_model)

model = model.to("cuda")
model.eval()
```

> Once loaded, you can pass inputs to tokenizer to prepare them, and call `model.generate()` in regular PyTorch fashion.

Reviewer 2: ```suggestion
Once loaded, you can pass your inputs to the tokenizer to prepare them, and call `model.generate()` in regular `transformers` fashion.
```

3. Beat in 1 egg and 1 teaspoon vanilla extract.
4. Mix in 1 1/4 cups all-purpose flour.
5. Stir in 1/2 teaspoon baking soda and 1/2 teaspoon salt.
6. Fold in 3/4 cup semisweet chocolate chips.
7. Drop by
```

If you want to load a specific PEFT model, you can click `Use in PEFT` in the model card and you will be given a working snippet!

<div class="flex justify-center">
<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/peft_repo_light_new.png"/>
<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/peft_repo.png"/>
</div>
<div class="flex justify-center">
<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/peft_snippet_light.png"/>
<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/hub/peft_snippet.png"/>
</div>

## Additional resources

* PEFT [repository](https://github.com/huggingface/peft)

> * PEFT [docs](https://huggingface.co/docs/peft/index)

Reviewer 2: ```suggestion
* PEFT [docs](https://huggingface.co/docs/peft/index)
* PEFT [models](https://huggingface.co/models?library=peft&sort=trending)
```
